R/join_keys.R

```diff
@@ -0,0 +1,599 @@
+#' join_keys(x)["ADSL", "ADRS"]
+mutate_join_keys.teal_data <- function(x, dataset_1, dataset_2, value) { # nolint
+  join_keys(x) <- mutate_join_keys(join_keys(x), dataset_1, dataset_2, value)
+  join_keys(x)
+}
+
+#' @rdname split_join_keys
+#' @export
+split_join_keys <- function(join_keys_obj) {
+  UseMethod("split_join_keys", join_keys_obj)
+}
+
+#' @rdname split_join_keys
+#' @export
+split_join_keys.default <- function(join_keys_obj) {
+  split_join_keys(join_keys(join_keys_obj))
+}
+
+#' Split the `JoinKeys` object into a named list of join keys objects with an element for each dataset
+#'
+#' @param join_keys_obj (`JoinKeys`) base object to get the keys from.
+#'
+#' @return (`list`) a list of `JoinKeys` object
+#'
+#' @rdname split_join_keys
+#'
+#' @export
+#'
+#' @examples
+#' jk <- join_keys()
+#' jk["ds1", "ds2"] <- "some_col"
+#' jk["ds1", "ds3"] <- "new_col"
+#' split_join_keys(jk)
+split_join_keys.JoinKeys <- function(join_keys_obj) {
+  assert_join_keys(join_keys_obj)
+
+  list_of_list_of_join_key_set <- lapply(
+    names(join_keys_obj),
+    function(dataset_1) {
+      lapply(
+        names(join_keys_obj[[dataset_1]]),
+        function(dataset_2) join_key(dataset_1, dataset_2, get_join_key(join_keys_obj, dataset_1, dataset_2))
+      )
+    }
+  )
+  res <- lapply(list_of_list_of_join_key_set, function(.x) do.call(join_keys, .x))
+  names(res) <- names(join_keys_obj)
+
+  logger::log_trace("JoinKeys keys split.")
+  return(res)
+}
+
+#' @rdname merge_join_keys
+#' @export
+merge_join_keys <- function(join_keys_obj, new_join_keys) {
+  UseMethod("merge_join_keys", join_keys_obj)
+}
+
+#' @rdname merge_join_keys
+#' @export
+merge_join_keys.default <- function(join_keys_obj, new_join_keys) {
+  merge_join_keys(join_keys(join_keys_obj), new_join_keys)
+}
+
+#' Merging a list (or one) of `JoinKeys` objects into the current `JoinKeys` object
+#'
+#' @rdname merge_join_keys
+#'
+#' @param join_keys_obj (`JoinKeys`) object to merge the new_join_keys.
+#' @param new_join_keys  `list` of `JoinKeys` objects or single `JoinKeys` object
+#'
+#' @return a new `JoinKeys` object with the resulting merge.
+#'
+#' @export
+#'
+#' @examples
+#' jk1 <- join_keys()
+#' jk1["ds1", "ds2"] <- "some_col"
+#'
+#' jk2 <- join_keys()
+#' jk2["ds1", "ds3"] <- "new_col"
+#'
+#' merge_join_keys(jk1, jk2)
+merge_join_keys.JoinKeys <- function(join_keys_obj, new_join_keys) {
+  assert_join_keys(join_keys_obj)
+
+  if (test_join_keys(new_join_keys)) {
+    new_join_keys <- list(new_join_keys)
+  }
+
+  checkmate::assert_list(new_join_keys, types = c("JoinKeys"), min.len = 1)
+
+  result <- join_keys_obj
+
+  for (jk in new_join_keys) {
+    for (dataset_1 in names(jk)) {
+      for (dataset_2 in names(jk[[dataset_1]])) {
+        result[[dataset_1]][[dataset_2]] <- jk[[dataset_1]][[dataset_2]]
+      }
+    }
+  }
+  logger::log_trace("JoinKeys keys merged.")
+  return(result)
+}
```

Comment:
Not sure about `new_join_keys` being a list of keys. `base::merge` has `x` and `y` and if somebody has multiple data.frames, one should use reduce. Let's keep it simple

```suggestion
merge_join_keys.JoinKeys <- function(join_keys_obj, new_join_keys) {
  assert_join_keys(join_keys_obj)
  assert_join_keys_alike(new_join_keys) # should be named and contain lists which should contain vectors, each list should be named
  
  modifyList(join_keys_obj, new_join_keys) # or reverse
  return(result)
 }
```

Reply:
Done!